synapseclient/client.py

```diff
@@ -1134,13 +1135,199 @@ def _getEntityBundle(self, entity, version=None, bitFlags=0x800 | 0x400 | 0x2 |
+
+        :param entity:          A synapse entity ID
+
+        :param destinationId:   Synapse ID of a folder/project that the copied entity is being copied to
+
+        :param copyWiki:        Determines whether the wiki of the entity is copied over
+                                Default is True
+
+        :param onlyCopyWiki:    Determines whether only the wiki of the entity is copied
+                                Default is False
+
+        :param copyTable:       Determines whether tables are copied
+                                Default is True
+
+        :param kwargs:          Parameters that can be passed to the hidden copy functions called
+        """
+        mapping = kwargs.get('mapping', dict())
+        updateLinks = kwargs.get('updateLinks', True)
+        updateSynIds = kwargs.get('updateSynIds', True)
+        if onlyCopyWiki:
+            self._copyWiki(entity, destinationId, updateLinks=updateLinks, updateSynIds=updateSynIds, entityMap=mapping)
+        else:
+            mapping = self._copyRecursive(entity, destinationId, mapping=mapping,**kwargs)
+            if copyWiki:
+                for oldEnt in mapping:
+                    newWikig = self._copyWiki(oldEnt, mapping[oldEnt], updateLinks=updateLinks, updateSynIds=updateSynIds, entityMap=mapping)
+        return(mapping)
+             
+    def _copyRecursive(self, entity, destinationId, mapping = dict(), **kwargs):
+        """
+        Recursively copies synapse entites, but does not copy the wikis
+
+        :param entity:             A synapse entity ID
+
+        :param destinationId:      Synapse ID of a folder/project that the copied entity is being copied to
+
+        :param mapping:            Takes a mapping {'oldSynId': 'newSynId'} to help with copying syn ids that aren't already part of the entity
+                                   Default to dict() and builds the mapping of all the synapse entities copied
+        """
+
+        version = kwargs.get('version', None)
+        setProvenance = kwargs.get('setProvenance', "traceback")
+        recursive = kwargs.get('recursive',True)
+        copyTable = kwargs.get('copyTable',True)
+        copiedId = None
+        ent = self.get(entity,downloadFile=False)
+        if isinstance(ent, Project):
+            if destinationId == None:
+                newProject = self.store(Project("Copied %s %s" %(time.time(),ent.name)))
+                destinationId = newProject.id
+            elif not isinstance(self.get(destinationId),Project):
+                raise ValueError("You must give a destinationId of a new project to copy projects")
+            mapping[ent.id] = destinationId
+            entities = self.chunkedQuery('select id, name from entity where parentId=="%s"' % ent.id)
+            for i in entities:
+                mapping = self._copyRecursive(i['entity.id'], destinationId, mapping=mapping, **kwargs)
+        elif isinstance(ent, Folder):
```

Comment:
Friendly error msg for the case where destinationId is None and we're not copying a project